Generics refactoring (groundwork for const generics)

jplatte opened the pull request:
These changes were suggested by @eddyb.

After this change, the `Generics` contain one `Vec` of an enum for the generic parameters, rather than two separate `Vec`s for lifetime and type parameters. Type params and const params will need to be in a shared `Vec` to preserve their ordering, and moving lifetimes into the same `Vec` should simplify the code that processes `Generics`.

eddyb commented on src/libsyntax_ext/deriving/generic/mod.rs:
```diff
@@ -598,7 +598,7 @@ impl<'a> TraitDef<'a> {
                     let predicate = ast::WhereBoundPredicate {
                         span: self.span,
-                        bound_lifetimes: vec![],
+                        bound_generics: Default::default(),
```
Here, and in a few other places, you have `Default::default()`, I would prefer `Generics::default()` for clarity.

eddyb commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1127,13 +1143,30 @@ impl<'a, 'tcx> Clean<Generics> for (&'a ty::Generics,
     }
 }
 
+impl Default for Generics {
```
You can just derive `Default` instead of manually implementing it.

jplatte replied:
Huh, dunno why I missed that this is the same thing that the derive would generate.

eddyb commented on src/librustc/hir/intravisit.rs:
```diff
@@ -279,6 +279,12 @@ pub trait Visitor<'v> : Sized {
     fn visit_ty(&mut self, t: &'v Ty) {
         walk_ty(self, t)
     }
+    fn visit_ty_param(&mut self, t: &'v TyParam) {
```
This isn't actually used - I'd prefer if people went through `visit_generic_param` or `visit_generics` instead - that way, when const generics are added, they're harder to miss.

Also, have you considered removing `visit_lifetime_def`? I believe that was the reason I mentioned changing `Vec<LifetimeDef>` to `Generics`, because then all the cases can be handled in one place in all visitors.

jplatte replied:
Do you mean inlining just `visit_ty_param`, or `walk_ty_param` too?

petrochenkov commented on src/libsyntax/ast.rs:
```diff
@@ -368,8 +393,8 @@ pub enum WherePredicate {
-    /// Any lifetimes from a `for` binding
-    pub bound_lifetimes: Vec<LifetimeDef>,
+    /// Any generics from a `for` binding
+    pub bound_generics: Generics,
```
There may be reasons to change `Vec<LifetimeDef>` into `Vec<GenericParam>` in the future, but why `Generics`?
I'd rather just leave `Vec<LifetimeDef>` in place, this change doesn't help const generics in any way and permits too many things in AST that can't happen in actual code.

petrochenkov replied:
Same for `BareFnTy` and `PolyTraitRef`.

eddyb replied:
IIRC there are a bunch of places where `visit_lifetime_def` is implemented separately from type parameters (`visit_generics`), only because `visit_generics` wouldn't be called in those `Vec<LifetimeDef>` cases (see also https://github.com/rust-lang/rust/pull/45930#discussion_r150382826).

petrochenkov replied:
I'm not sure the current situation is significantly worse than a bunch of impossible "`span_bug`" code.

If `visit_lifetime_def` is removed and `visit_generic_param` is uniformly used for visiting (https://github.com/rust-lang/rust/pull/45930#discussion_r150382826) then `Generics` here could be replaced with `Vec<GenericParam>` improving situation somewhat.

eddyb replied:
FWIW I expect @nikomatsakis would eventually want HRTB to include `where` clauses too, on top of `for<T: Trait>`, but I don't know the exact plans.
`Vec<GenericParam>` is fine with me too.

petrochenkov commented on src/libsyntax/ast.rs:
```diff
@@ -297,30 +297,56 @@ pub struct TyParam {
 pub struct Generics {
-    pub lifetimes: Vec<LifetimeDef>,
-    pub ty_params: Vec<TyParam>,
+    pub params: Vec<GenericParam>,
```
Also the invariant "all lifetime parameters go before type parameters" is lost in the new AST/HIR (enforced only by parser).
I don't remember if it was significant anywhere, in matching against generic arguments maybe?

petrochenkov replied:
What happens if a syntax extension generates parameters in intermixed order and send them to further compilation?

eddyb replied:
`ast_validation` should probably enforce the limitation until we want to lift it.

bors commented:
:umbrella: The latest upstream changes (presumably #45848) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
r? @eddyb 

(just making sure there's an assigned reviewer)

bors commented:
:umbrella: The latest upstream changes (presumably #45701) made this pull request unmergeable. Please resolve the merge conflicts.

carols10cents commented:
@jplatte friendly triage ping! how's this going? Looks like there are some merge conflicts for you!

jplatte commented:
I think the test failures still are the bigger issue.. I've rebased a bunch of times and I will do that again now. But I've put some time into figuring out why some tests fail and haven't been able to figure it out yet. I have one idea still, but I'm not that optimistic...

eddyb commented on src/librustc/hir/lowering.rs:
```diff
@@ -1264,17 +1278,20 @@ impl<'a> LoweringContext<'a> {
+                                                if node_id == ty_param.id {
+                                                    add_bounds.entry(ty_param.id)
+                                                        .or_insert(Vec::new())
+                                                        .push(bound.clone());
```
FWIW cloning isn't necessary here if the reference can be used directly.

eddyb commented on src/librustc/hir/map/def_collector.rs:
```diff
@@ -190,13 +190,26 @@ impl<'a> visit::Visitor<'a> for DefCollector<'a> {
+    }
 
+    fn visit_generics(&mut self, generics: &'a Generics) {
         visit::walk_generics(self, generics);
```
This can be removed, since walking is the default action.

eddyb commented on src/librustc/hir/mod.rs:
```diff
@@ -403,15 +423,31 @@ impl Generics {
 
     pub fn is_parameterized(&self) -> bool {
-        self.is_lt_parameterized() || self.is_type_parameterized()
+        !self.params.is_empty()
```
Could maybe inline this, since it's about as long.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -340,15 +340,18 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
-                    this.check_lifetime_defs(old_scope, &generics.lifetimes);
+                    this.check_lifetime_defs(
+                        old_scope,
+                        &generics.lifetimes().cloned().collect::<Vec<_>>(),
```
Should avoid doing this cloning, preferably just pass in the `&[GenericParam]`.

jplatte replied:
`generics.lifetimes().collect::<Vec<_>>()[..]` is a `&[&GenericParam]` though, is there a way of converting that or would I have to make `check_lifetime_defs` generic over `Iterator<Item = &GenericParam>`?

eddyb replied:
I'd rename `check_lifetime_defs` to `check_lifetime_params` and have it take the entire `&[GenericParam]`, and use `.lifetimes()` inside itself.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -375,15 +378,32 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
-                        }).collect(),
+                    lifetimes: c.generic_params
+                        .iter()
+                        .flat_map(|param| if let hir::GenericParam::Lifetime(ref def) = *param {
```
Hmm, you're not using `.lifetimes()` here, but you are above, is that intentional?

jplatte replied:
Can't because this is a `Vec<GenericParam>`, not a `Generics` instance. In my WIP rebase, this case actually came up a bunch more and I added a `GenericParmsExt` to deal with this (where I had to write out the FlatMap type bc. impl trait in traits isn't available yet), but then had to realize that it's sometimes `hir::GenericParam` and sometimes `ast::GenericParam`.

eddyb replied:
At least in this file it's always HIR.

jplatte replied:
I just re-checked and it's actually almost always HIR, just in the form of `P<[GenericParam]>` in some cases. Only very few things that were new when I started the rebase were in the AST (in the ast -> hir lowering specifically).

eddyb replied:
You can impl the trait on `[GenericParam]` directly.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -375,15 +378,32 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
+                        &c.generic_params
+                            .iter()
+                            .flat_map(|param| {
+                                if let hir::GenericParam::Lifetime(ref def) = *param {
```
Same as https://github.com/rust-lang/rust/pull/45930#discussion_r153937804.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -463,28 +483,46 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
+                    ref bound_generic_params,
+                    ..
+                }) => {
+                    if bound_generic_params.iter().find(|p| p.is_lifetime_param()).is_some() {
```
This is just `.any(|p| p.is_lifetime_param())`, right?

jplatte replied:
Yeah

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -463,28 +483,46 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
-                                }).collect(),
+                            lifetimes: bound_generic_params.iter()
+                                .flat_map(|param| {
+                                    if let hir::GenericParam::Lifetime(ref def) = *param {
```
Same as https://github.com/rust-lang/rust/pull/45930#discussion_r153937946.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -463,28 +483,46 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
+                                old_scope,
+                                &bound_generic_params.iter()
+                                    .flat_map(|param| {
+                                        if let hir::GenericParam::Lifetime(ref ld) = *param {
```
Same as https://github.com/rust-lang/rust/pull/45930#discussion_r153937804.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -519,22 +557,38 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
 
-        if !self.trait_ref_hack || !trait_ref.bound_lifetimes.is_empty() {
+        if !self.trait_ref_hack ||
+            trait_ref.bound_generic_params.iter().find(|p| p.is_lifetime_param()).is_some()
```
Same as https://github.com/rust-lang/rust/pull/45930#discussion_r153938396.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -519,22 +557,38 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
-                    }).collect(),
+                lifetimes: trait_ref.bound_generic_params.iter()
+                    .flat_map(|param| {
+                        if let hir::GenericParam::Lifetime(ref def) = *param {
```
Same as https://github.com/rust-lang/rust/pull/45930#discussion_r153937946.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -519,22 +557,38 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
+                    old_scope,
+                    &trait_ref.bound_generic_params.iter()
+                        .flat_map(|param| {
+                            if let hir::GenericParam::Lifetime(ref ld) = *param {
```
Same as https://github.com/rust-lang/rust/pull/45930#discussion_r153937804.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -846,26 +906,31 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
-            this.check_lifetime_defs(old_scope, &generics.lifetimes);
+            this.check_lifetime_defs(
+                old_scope,
+                &generics.lifetimes().cloned().collect::<Vec<_>>(),
```
Same as https://github.com/rust-lang/rust/pull/45930#discussion_r153937804.

eddyb commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -1246,6 +1311,12 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
+            fn visit_generic_param(&mut self, param: &hir::GenericParam) {
+                if let hir::GenericParam::Lifetime(ref lifetime_def) = *param {
+                    for l in &lifetime_def.bounds { self.visit_lifetime(l); }
+                }
```
You're not walking over the parameter if it's anything else!

eddyb commented on src/librustc_resolve/lib.rs:
```diff
@@ -766,25 +766,30 @@ impl<'a, 'tcx> Visitor<'tcx> for Resolver<'a> {
         let mut default_ban_rib = Rib::new(ForwardTyParamBanRibKind);
-        default_ban_rib.bindings.extend(generics.ty_params.iter()
+        default_ban_rib.bindings.extend(generics.params.iter()
+            .flat_map(|p| if let GenericParam::Type(ref tp) = *p { Some(tp) } else { None })
```
Shouldn't this use `.ty_params()`? Although, in the long term, we should be moving named lifetime resolution to this pass.

jplatte replied:
I only added `lifetimes()` and `ty_params()` to hir, not libsyntax, given that the former is part of the rustc crate that already uses `conservative_impl_trait` and the latter doesn't currently use any nightly features.

eddyb replied:
Ah that's fine then.

eddyb commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -4998,24 +4998,24 @@ pub fn check_bounds_are_used<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
         if let ty::TyParam(ParamTy {idx, ..}) = leaf_ty.sty {
             debug!("Found use of ty param num {}", idx);
-            tps_used[idx as usize - generics.lifetimes.len()] = true;
+            tps_used[idx as usize - generics.lifetimes().count()] = true;
```
You should at least precompute this instead of having it in the loop.

eddyb commented on src/libsyntax/feature_gate.rs:
```diff
@@ -1628,22 +1628,33 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
+    fn visit_generics(&mut self, g: &'a ast::Generics) {
+        for param in &g.params {
+            match *param {
+                ast::GenericParam::Type(ref t) => {
```
This can be moved into `visit_generic_param`.

eddyb commented on src/libsyntax/parse/parser.rs:
```diff
@@ -5231,16 +5230,24 @@ impl<'a> Parser<'a> {
-                              "only lifetime parameters can be used in this context");
+
+            let first_non_lifetime_param_span = params.iter()
+                .flat_map(|param| match *param {
```
This should be `filter_map`.

eddyb commented on src/librustc/hir/mod.rs:
```diff
@@ -403,15 +423,31 @@ impl Generics {
+
+    pub fn ty_params<'a>(&'a self) -> impl Iterator<Item = &'a TyParam> {
+        self.params.iter()
+            .flat_map(|param| match *param {
```
Same here, this and `lifetimes` above should use `filter_map`.

eddyb commented on src/librustc/hir/map/def_collector.rs:
```diff
@@ -190,13 +190,26 @@ impl<'a> visit::Visitor<'a> for DefCollector<'a> {
+                    REGULAR_SPACE
+                );
+            }
         }
```
You forgot a `walk_generic_param` here.

jplatte commented:
I addressed all of the latest feedback and rebased again. I still need to add the ast validation pass for the generic parameter ordering (enforce no type params before lifetime params) but I won't be able to do that today.

bors commented:
:umbrella: The latest upstream changes (presumably #46657) made this pull request unmergeable. Please resolve the merge conflicts.

jplatte commented:
Rebased, fixed the tidy errors, got rid of the remaining `flat_map`s and added the AST validation :)

eddyb commented on _config.toml:
```diff
@@ -0,0 +1,385 @@
+# Sample TOML configuration file for building Rust.
```
Committing this seems accidental.

jplatte replied:
Yup. Will remove it.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 573e094 has been approved by `eddyb`

<!-- @bors r=eddyb 573e094c58531b8448f43488326f913785f6dbdb -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 2, this pull request will be tested once the tree is reopened

kennytm commented:
@bors r-

<details><summary>CI is failing (run-make/pretty-print-path-suffix).</summary>

```
[01:02:57] ---- [run-make] run-make/pretty-print-path-suffix stdout ----
[01:02:57] 	
[01:02:57] error: make failed
[01:02:57] status: exit code: 2
[01:02:57] command: "make"
[01:02:57] stdout:
[01:02:57] ------------------------------------------
[01:02:57] make[1]: Entering directory '/checkout/src/test/run-make/pretty-print-path-suffix'
[01:02:57] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib:" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu  -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu/foo.out -Z unstable-options --unpretty hir=foo input.rs
[01:02:57] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib:" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu  -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu/nest_foo.out -Z unstable-options --unpretty hir=nest::foo input.rs
[01:02:57] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib:" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu  -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu/foo_method.out -Z unstable-options --unpretty hir=foo_method input.rs
[01:02:57] diff -u /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu/foo.out foo.pp
[01:02:57] --- /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/pretty-print-path-suffix.stage2-x86_64-unknown-linux-gnu/foo.out	2017-12-13 17:46:32.935556154 +0000
[01:02:57] +++ foo.pp	2017-12-13 16:42:30.499819848 +0000
[01:02:57] @@ -9,7 +9,7 @@
[01:02:57]  // except according to those terms.
[01:02:57]  
[01:02:57]  
[01:02:57] -pub fn foo<>() -> i32 { 45 } /* foo */
[01:02:57] +pub fn foo() -> i32 { 45 } /* foo */
[01:02:57]  
[01:02:57]  
[01:02:57] -pub fn foo<>() -> &'static str { "i am a foo." } /* nest::foo */
[01:02:57] +pub fn foo() -> &'static str { "i am a foo." } /* nest::foo */
[01:02:57] Makefile:4: recipe for target 'all' failed
[01:02:57] make[1]: Leaving directory '/checkout/src/test/run-make/pretty-print-path-suffix'
[01:02:57] 
[01:02:57] ------------------------------------------
[01:02:57] stderr:
[01:02:57] ------------------------------------------
[01:02:57] make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[01:02:57] warning: ignoring --out-dir flag due to -o flag.
[01:02:57] 
[01:02:57] warning: ignoring --out-dir flag due to -o flag.
[01:02:57] 
[01:02:57] warning: ignoring --out-dir flag due to -o flag.
[01:02:57] 
[01:02:57] make[1]: *** [all] Error 1
[01:02:57] 
[01:02:57] ------------------------------------------
[01:02:57] 
[01:02:57] thread '[run-make] run-make/pretty-print-path-suffix' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2774:8
[01:02:57] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:02:57] 
[01:02:57] 
[01:02:57] failures:
[01:02:57]     [run-make] run-make/pretty-print-path-suffix
```

</details>

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit ab5fb1d has been approved by `eddyb`

<!-- @bors r=eddyb ab5fb1deb5196891e7b868cc04e25b70ad33f8a8 -->

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 3119862 has been approved by `eddyb`

<!-- @bors r=eddyb 3119862b7c383defc4bb618210a32ebd909e5bb1 -->

bors commented:
:hourglass: Testing commit 3119862b7c383defc4bb618210a32ebd909e5bb1 with merge e790ba700f95705b293f95e80baae194cb115c9d...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/316229452?utm_source=github_status&utm_medium=notification)

kennytm commented:
Rustfmt and rls are broken.

bors commented:
:umbrella: The latest upstream changes (presumably #46778) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit b43275f has been approved by `eddyb`

<!-- @bors r=eddyb b43275f9081821b03058688c683a1f4383cfcec9 -->

bors commented:
:hourglass: Testing commit b43275f9081821b03058688c683a1f4383cfcec9 with merge 92c7dd4b58c9cdd56633015888cb25284ab1496a...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/319498417?utm_source=github_status&utm_medium=notification)

eddyb commented:
I think you need to rebase, because some NLL testcases got added just now and they break.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit b9e4b30 has been approved by `eddyb`

<!-- @bors r=eddyb b9e4b301c59976230dce67f1ab8eae79b8905db3 -->

bors commented:
:lock: Merge conflict

bors commented:
:umbrella: The latest upstream changes (presumably #46754) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 78493ed has been approved by `eddyb`

<!-- @bors r=eddyb 78493ed21aaeb20a8d5f7bd08f26c0c9fd496ed4 -->

bors commented:
:hourglass: Testing commit 78493ed21aaeb20a8d5f7bd08f26c0c9fd496ed4 with merge 250b492052e94d76bd318a4b932159395da55981...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5775), [status-travis](https://travis-ci.org/rust-lang/rust/builds/319890498?utm_source=github_status&utm_medium=notification)
Approved by: eddyb
Pushing 250b492052e94d76bd318a4b932159395da55981 to master...
